ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,862 @@
+            .filter(|hint| matches!(hint, Hint::CommitmentHint(_)))
+            .collect();
+        let secret: Vec<Hint> = hints_bag
+            .hints
+            .into_iter()
+            .filter(|hint| matches!(hint, Hint::SecretProven(_)))
+            .collect();
+
+        self.secret_hints.insert(index, HintsBag { hints: secret });
+        self.public_hints.insert(index, HintsBag { hints: public });
+    }
+
+    /// Adding hints for a input index
+    pub fn add_hints_for_input(&mut self, index: usize, hints_bag: HintsBag) {
+        let mut public: Vec<Hint> = hints_bag
+            .hints
+            .clone()
+            .into_iter()
+            .filter(|hint| matches!(hint, Hint::CommitmentHint(_)))
+            .collect();
+        let mut secret: Vec<Hint> = hints_bag
+            .hints
+            .into_iter()
+            .filter(|hint| matches!(hint, Hint::SecretProven(_)))
+            .collect();
+        let secret_bag = HintsBag::empty();
+        let public_bag = HintsBag::empty();
+        let old_secret: &Vec<Hint> = &self.secret_hints.get(&index).unwrap_or(&secret_bag).hints;
+        for hint in old_secret {
+            secret.push(hint.clone());
+        }
+
+        let old_public: &Vec<Hint> = &self.public_hints.get(&index).unwrap_or(&public_bag).hints;
+        for hint in old_public {
+            public.push(hint.clone());
+        }
+        self.secret_hints.insert(index, HintsBag { hints: secret });
+        self.public_hints.insert(index, HintsBag { hints: public });
+    }
+
+    /// Outputting HintsBag corresponding for an index
+    pub fn all_hints_for_input(&self, index: usize) -> HintsBag {
+        let mut hints: Vec<Hint> = Vec::new();
+        let secret_bag = HintsBag::empty();
+        let public_bag = HintsBag::empty();
+        let secrets: &Vec<Hint> = &self.secret_hints.get(&index).unwrap_or(&secret_bag).hints;
+        for hint in secrets {
+            hints.push(hint.clone());
+        }
+        let public: &Vec<Hint> = &self.public_hints.get(&index).unwrap_or(&public_bag).hints;
+        for hint in public {
+            hints.push(hint.clone());
+        }
+        let hints_bag: HintsBag = HintsBag { hints };
+        hints_bag
+    }
+}
+
+/// Compute commitments for UncheckedLeaf
+pub fn compute_commitments(leaf: UncheckedLeaf) -> Option<FirstDlogProverMessage> {
+    let mut ret: Option<FirstDlogProverMessage> = None;
+
+    match leaf {
+        UncheckedLeaf::UncheckedSchnorr(pdlog) => {
+            let challenge = pdlog.challenge;
+            let proposition = pdlog.proposition;
+            let second_message = pdlog.second_message;
+            let comm = compute_commitment(&proposition, &challenge, &second_message);
+            ret = Some(FirstDlogProverMessage::from(comm));
+        }
+        UncheckedLeaf::UncheckedDhTuple(_) => {}
+    }
+    ret
+}
+
+#[allow(clippy::unwrap_used)]
+/// Outputs Hints bag for a signed(invalid) transaction
```

Comment:
I believe doc comment could be better. We can copy from https://github.com/ergoplatform/ergo/blob/c320810c498bca25a44197840c7c5a86440c5906/.metals/readonly/dependencies/sigma-state_2.12-4.0.3-sources.jar/sigmastate/interpreter/ProverUtils.scala#L64-L76